Typescript conversion

blms opened the pull request:
## What this PR does

- Converts all JS to TypeScript
- Refactors HTML elements into ES6 class modules that are implemented as custom elements, and uses the main `TranscriptionEditor` class primarily to perform CRUD-like operations on the annotations.
- Adds documentation of classes, methods, and types with JSDoc

rlskoeser reviewed (commented):
What you have so far looks good to me.

Should type definitions have comments or docstrings?

Are there any shared types between this and the storage plugin?

I was hoping to keep the two plugins somewhat loosely coupled, but given that this we don't (at this point) want to worry about making this one generalizable and plan for it to be pretty specific to PGP, it seems ok to depend on the sas storage plugin if that makes things easier.

blms commented:
Just pushed an update where I moved the save/cancel/delete buttons into their own ES modules. This isn't strictly necessary or anything, but I think it makes the main index code more readable. Plus, it was fun to work with custom elements! (Basically followed the design pattern of React Components but using custom elements instead.)

> Should type definitions have comments or docstrings?

Yes; I think a good task for early next week would be looking into best practices for TS and JS documentation, and applying those to both projects. It shouldn't take much time. I'll work on that before re-requesting review!

> Are there any shared types between this and the storage plugin?
>
> I was hoping to keep the two plugins somewhat loosely coupled, but given that this we don't (at this point) want to worry about making this one generalizable and plan for it to be pretty specific to PGP, it seems ok to depend on the sas storage plugin if that makes things easier.

Yes, we should come up with a solution for sharing them. Makes sense to me to rely on exported types from the storage plugin; I'll take a look next week.

blms commented:
Progress update:

- Added JSDoc documentation. I added a _lot_ of this, maybe more than we want. There was actually one step further I could have taken it, which is documenting each property of each TS type interface!
  - I also set up ESLint to auto-generate empty documentation for new classes, methods, and types. It will raise warnings when you don't write descriptions for them, and auto-correct any formatting errors.
  - There are tools out there to generate static site documentation from these comments, if we want to use them.
- Noticed there was inconsistent naming around "editor block"/"display block"/"container". I ended up calling them "display blocks" since that seemed to be the most used, but I'm not sure how good of a term it is.
- Refactored the buttons a bit. I'm not totally satisfied with the way I handled them, identical data seems to get passed around too many times and it got really verbose. I think defining a class for `DisplayBlock` (or whatever we call it) would help; a lot of the idiosyncrasies have to do with making those read-only or deleting them.

Didn't manage to get to testing yet today, I can pick that up tomorrow. In the meantime it might make sense to get another review on this!

rlskoeser commented:
@blms the js documentation looks great, thank you for adding this.

Seeing the separate modules for each button is reminding me why I sometimes dislike javascript! 😂 But this approach does make sense to me.

I think you're right that a class for the display block would help substantially with making the code more organized and easier to work with — I'm with you on how things are getting passed around right now. The class could then encapsulate logic for displaying/suppressing buttons, switching between edit and read only mode; and I think that's where we'd want to turn off and on the tinymce integration when we add it. 

Any thoughts on names that would be more meaningful? `AnnotationBlock` maybe?

blms commented:
Ended up going with `AnnotationBlock`, I like it! Refactored to move that into a separate module, and now I think it's much cleaner. 

There two slightly ugly parts are the _ton_ of parameters I'm passing to the `AnnotationBlock` constructor, and the fact that the `AnnotationBlock` sometimes passes itself to functions (e.g. `this.annotationBlock.onSave(this.annotationBlock)`), but I think it's still a big improvement. 

To alleviate the former issue, I passed an object with the parameters as a single argument, instead of each one individually, so that they can act like kwargs in Python. There might be a better pattern for the latter issue… I could experiment a bit.

Once we can approve the changes so far, does it make sense to merge this into `develop` and start a separate branch for tests? That way I would feel comfortable committing the changes to our main geniza transcription-editor branch where we pull this library from `develop`.

blms commented:
Discovered a bug here that seems to be present in the original code too, but not sure how to fix it: if you change the Annotorious selection and save it, then edit the text of the corresponding annotation and save it again, the Annotorious selection reverts to its original shape/size.

rlskoeser commented on src/index.ts:
```diff
@@ -0,0 +1,213 @@
+     *
+     * @param {Annotation} selection Selected Annotorious annotation.
+     */
+    async handleCreateSelection(selection: Annotation) {
```
technically this returns a selection, not a full annotation https://recogito.github.io/annotorious/api-docs/osd-plugin/#createselection

I'm not sure how much the difference matters for us, and creating new annotations does seem to be working, but I noticed this when I was looking into the problem you identified.

blms replied:
I noticed that nomenclature too… I don't know the whole W3C Web Annotation spec, but I wasn't able to find anything with `"type": "Selection"` as Annotorious sets it. And other than the `type` field it seems to conform to the spec of an Annotation ([example](https://www.w3.org/TR/annotation-model/#sets-of-bodies-and-targets)). (I did find [this](https://www.w3.org/TR/selection-api/#selection-interface) but it seems unrelated.)

rlskoeser commented on src/index.ts:
```diff
@@ -0,0 +1,213 @@
+                new AnnotationBlock({
+                    annotation,
+                    editable: false,
+                    onCancel: this.anno.cancelSelected,
+                    onClick: this.handleClickAnnotationBlock.bind(this),
+                    onDelete: this.handleDeleteAnnotation.bind(this),
+                    onSave: this.handleSaveAnnotation.bind(this),
+                    updateAnnotorious: this.anno.addAnnotation,
```
I agree with you about disliking how we're binding things currently. 

Can we resolve this by approaching the way we're supposed to for Turbo anyway, by using event delegation and binding once on document or window? Discussed briefly here https://turbo.hotwired.dev/handbook/building#observing-navigation-events 

Let me know if this isn't clear. I'm not sure we need to address now, unless it solves a problem.

blms replied:
That makes sense to me—put the event listeners on the entire transcription editor and not just each block. I agree it's not necessary now, more of a cleanup chore.

blms replied:
FWIW, these events are being attached after the `document` event `annotations-loaded` is fired, so they're meeting the Turbo expectation to some extent, but we are breaking it in other places.

rlskoeser commented on src/elements/AnnotationBlock.ts:
```diff
@@ -0,0 +1,142 @@
+ * or delete its associated annotation.
+ */
+class AnnotationBlock extends HTMLDivElement {
+    annotation: Annotation;
```
I'm wondering if the cached annotation here is the cause of the problem you noticed with saving changes to the selection. It seems like it would be better to use the annotorious method to get annotation by id when needed.

I've experimented with it some and I think the changed selection target isn't getting stored. There's a change selection target event we could listen to, but I'm not sure that we should (it fires repeatedly when dragging to adjust).
https://recogito.github.io/annotorious/api-docs/osd-plugin/#changeselectiontarget

blms replied:
That's helpful, thanks! I agree, let's file it as an issue (maybe on the geniza repo to keep things tidy) and merge this in without fixing.

rlskoeser reviewed (commented):
I'm ok with merging into develop and doing subsequent rounds of improvements. 

My inclination at this point is to not worry about the bug you've identified — we should make a note of it somewhere, but not sink more time into fixing ut. There are almost certainly other problems, but I think this code needs a bigger step back to plan a systematic approach, which should be easier once Gissoo's designs are finalized.

Your approach to bundle params as a single object similar to python kwargs is one I've seen and used myself before; I like this pattern. 

It seems like the unit testing we've done so far may be too brittle or too tightly coupled (or maybe it's the mocks) — I don't want to spend a lot of time on writing and rewriting testing, when we know this code is preliminary and likely to change. Let's just do some basic test scaffolding without testing exhaustively, for now (but can be a separate pr).
